pkg/service/gcp/volume.go

```diff
@@ -0,0 +1,147 @@
+package gcp
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"gitlab.com/netbook-devs/spawner-service/pkg/service/labels"
+	"gitlab.com/netbook-devs/spawner-service/pkg/service/system"
+	proto "gitlab.com/netbook-devs/spawner-service/proto/netbookai/spawner"
+	disk_proto "google.golang.org/genproto/googleapis/cloud/compute/v1"
+)
+
+func diskName(size int32) string {
+	t := time.Now().Format("20060102150405")
+
+	return fmt.Sprintf("vol-%d-%s", size, t)
+}
+
+func diskType(projectId, zone, typ string) string {
+	return fmt.Sprintf("projects/%s/zones/%s/diskTypes/%s", projectId, zone, typ)
+}
+
+func (g *GCPController) createVolume(ctx context.Context, req *proto.CreateVolumeRequest) (*proto.CreateVolumeResponse, error) {
+	cred, err := getCredentials(ctx, req.AccountName)
+	if err != nil {
+		return nil, errors.Wrap(err, "createVolume ")
+	}
+
+	client, err := getDiskClient(ctx, cred)
+	if err != nil {
+		g.logger.Error(ctx, "failed to get disk client", "error", err)
+		return nil, errors.Wrap(err, "failed to get disk client")
+	}
+	//FIXME: hard coded zone info
+	zone := fmt.Sprintf("%s-a", req.Region)
+	diskType := diskType(cred.ProjectId, zone, "pd-balanced")
+	size := req.Size
+	name := diskName(int32(size))
+	label := make(map[string]string)
+	for k, v := range labels.DefaultTags() {
+		label[k] = *v
+	}
+
+	for k, v := range req.Labels {
+		label[k] = v
+
+	}
+	idr := disk_proto.InsertDiskRequest{
+		DiskResource: &disk_proto.Disk{
+			Labels: label,
+			Name:   &name,
+			SizeGb: &size,
+			Type:   &diskType,
+		},
+		Project: cred.ProjectId,
+		Zone:    zone,
+	}
+
+	if req.Snapshotid != "" {
+
+		g.logger.Info(ctx, "creating disk from snapshot", "snapshot", req.Snapshotid)
+		source := fmt.Sprintf("global/snapshots/%s", req.Snapshotid)
+		idr.DiskResource.SourceSnapshot = &source
+	}
+	g.logger.Info(ctx, "creating volume", "name", name, "size", size)
+	// Doc : https://cloud.google.com/compute/docs/reference/rest/v1/disks/insert
+	r, err := client.Insert(ctx, &idr)
+	if err != nil {
+		g.logger.Error(ctx, "failed to insert/create new disk", "error", err)
+		return nil, errors.Wrap(err, "disk.Insert failed")
+
+	}
+	err = r.Wait(ctx)
+	if err != nil {
+		g.logger.Error(ctx, "wait on volume create failed")
+		return nil, errors.Wrap(err, "create volume wait failed")
+	}
+
+	g.logger.Info(ctx, "disk created", "name", name, "zone", zone)
+
+	if req.DeleteSnapshot {
+		go func() {
+
+			deleteSnapshotTimeout := time.Duration(10) * time.Second
```

Comment:
Larger snapshots or network clogging might make snapshot deletion time longer, can we increase this to a couple of min?

Reply:
its the request time, not the actual delete time. Same as aws.